File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
#include <string>

using namespace std;


int main(void)
{
	int a;
	cin >> a;
	while (a--)
	{
		string arr;
		stack<char> s;
		bool flag = false;
		cin >> arr;
		for (char b : arr)
		{
			if (b == '(')
			{
				s.push('(');
			}
			else if (s.empty())
			{
				flag = true;
				cout << "NO" << endl;
				break;
			}
			else s.pop();
		}
		if (!flag)
		{
			if (s.empty()) cout << "YES" << endl;
			else cout << "NO" << endl;
		}
	}
	return 0;
}

File: 백준/Silver/9012. 괄호/괄호.cs
using System;
using System.Collections.Generic;
using System.Text;

public class BOJ9012
{
    public static void Main()
    {
        int N = int.Parse(Console.ReadLine());
        var sb = new StringBuilder();
        
        for (int i = 0; i < N; i++)
        {
            string s = Console.ReadLine();
            var st = new Stack<char>();
            bool isVPS = true;
            foreach (char c in s)
            {
                if (c == '(')
                {
                    st.Push(c);
                }
                
                else if (c == ')')
                {
                    if (st.Count > 0) st.Pop();
                    else
                    {
                        isVPS = false;
                        break;
                    }
                }
                
            }
            if (st.Count == 0 && isVPS) sb.AppendLine("YES");
            else sb.AppendLine("NO");
        }
        Console.Write(sb.ToString());
    }
}
